[3.1] [WebProfilerBundle] [DX] Feature allow forward and redirection detection in wdt

HeahDude opened the pull request:
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #14358, #17501 |
| License | MIT |
| Doc PR | ? |

This PR allows to :
- track explicit forward from `\Symfony\Bundle\FrameWorkBundle\Controller\Controller` in the web debug toolbar.
- or pass a request attribute `_forwarded` with the current request attributes (an instance of `ParameterBag`) as value to your sub request before handling it.
- see if you've been redirected (require session enabled)

When redirected you will see the name of the route (if any) and a link to the profile of the original request.

![redirect](https://cloud.githubusercontent.com/assets/10107633/12716952/9aacdcba-c8e4-11e5-9a64-d26fe27f1cae.jpg)

In case of forwarding, the name of the controller is a file link and next to it there is a direct link to the profile of the sub request.

![forward](https://cloud.githubusercontent.com/assets/10107633/12716968/ba6b1fbc-c8e4-11e5-85fc-7f71969cb372.jpg)

This works pretty well in **Silex** too by registering `SessionServiceProvider()` for redirections or by providing this method for forwarding :

``` php
class App extends \Silex\Application
//  (php7 bootstrap) $app = new class extends \Silex\Application {
{
    public function forward($controller, array $path = array(), array $query = array()
    {
        if (!$this->booted) {
            throw new LogicException(sprintf('Method %s must be called from a controller.', __METHOD__));
        }

        $this->flush();

        $request = $this['request_stack']->getCurrentRequest();
        $path['_forwarded'] = $request->attributes;
        $path['_controller'] = $controller;
        $subRequest = $request->duplicate($query, null, $path);

        return $this['kernel']->handle($subRequest, HttpKernelInterface::SUB_REQUEST);
    }
}
```

HeahDude committed:
[WebProfiler] [HttpKernel] profile redirections

closes #17501.

The profiler stores the current request attributes in a
current session when a `RedirectionResponse` is returned.
So the next request profile will inherit the previous request
attributes.

The main profiler layout displays a shortcut to a previous
redirection profile, along with some useful informations.

The web debug toolbar shows a notifying icon, meaning a shortcut
to a redirection profile is available in the request toolbar panel.

javiereguiluz commented on src/Symfony/Component/HttpKernel/DataCollector/RequestDataCollector.php:
```diff
@@ -276,18 +250,37 @@ public function getRouteParams()
+     * Gets the parsed forward controller.
+     *
+     * @return array|bool The parsed forward controller as an array of data
+     *                    with keys 'token' the forward profile token, and
```
It's nice to have good comments but I think you are revealing too many implementation details in this one.

javiereguiluz commented on src/Symfony/Component/HttpKernel/DataCollector/RequestDataCollector.php:
```diff
@@ -339,4 +332,60 @@ private function getCookieHeader($name, $value, $expires, $path, $domain, $secur
+     * @param $controller
+     *
+     * @return array|string An array of controller data with keys :
+     *                      'class', 'method', 'file' and 'line,
```
Same here. This comment contains too many implementation details.

HeahDude replied:
alright but previous one was wrong anyway it needs a clearer description IMHO

javiereguiluz commented on src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/request.html.twig:
```diff
@@ -18,11 +18,31 @@
         {% endif %}
     {% endset %}
 
+    {% if collector.forward %}
```
What do you think about creating a simple Twig macro to avoid repeating in `forward_handler` the same logic as the previous `request_handler`?

HeahDude replied:
I thought about it but where should it live ?

javiereguiluz replied:
In the same template. We already do that in other panels. Example: https://github.com/symfony/symfony/blob/master/src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/logger.html.twig

stof commented on src/Symfony/Bundle/FrameworkBundle/Controller/Controller.php:
```diff
@@ -64,8 +64,10 @@ protected function generateUrl($route, $parameters = array(), $referenceType = U
     protected function forward($controller, array $path = array(), array $query = array())
     {
+        $request = $this->container->get('request_stack')->getCurrentRequest();
+        $path['_forwarded'] = $request;
```
I don't think passing the whole parent request is a good idea. It adds a bunch of unrelated info in the subrequest attributes.

HeahDude replied:
@stof It appears to me as the simplest way to keep the forward trace, fortunately it's just an object reference, any suggestion ? maybe pass only the attributes bag ?

stof commented on src/Symfony/Component/HttpKernel/DataCollector/RequestDataCollector.php:
```diff
@@ -122,48 +122,22 @@ public function collect(Request $request, Response $response, \Exception $except
+
+        if ($parentRequest = $request->attributes->get('_forwarded')) {
+            if ($parentRequest instanceof Request) {
+                $parentRequest->attributes->set('_forward_token', $response->headers->get('x-debug-token'));
```
technically, there is nothing preventing a request handling to use forward for multiple subrequests (even though I don't really see why someone would do it, but I just never used forward at all anyway)

HeahDude replied:
@stof I agree, but `forward` method should be IMO used only once for a global dynamic re-routing to a controller with the responsibility to send the main response, I don't see any use case for multiple calls. Anyway nested ones would simply show a wdt shortcut to sub request profile, the last of a chain.

HeahDude commented:
Made some changes to allow catch redirections as well, and use only the master request attributes (`ParameterBag`) instead of the whole request to handle forward interception as @stof suggested.
Also updated title and description.
Now need some reviews, thanks.

HeahDude commented:
I wonder if it would be useful to add the redirection code to quickly see if it's the same as expected

javiereguiluz commented:
The designer just sent me the icons which match the style of the toolbar:

`redirect.svg`

```xml
<svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"><path d="M23.06,7.83L14,0.38a1.25,1.25,0,0,0-2,.89V4.09a13.61,13.61,0,0,1-2.2.61l-1.3.47C8,5.35,7.59,5.6,7.12,5.81l-0.69.35-0.72.45a10.62,10.62,0,0,0-1.41,1A13.22,13.22,0,0,0,3,8.82a15.31,15.31,0,0,0-1.13,1.46A17.63,17.63,0,0,0,1,11.93c-0.18.58-.34,1.16-0.48,1.71S0.45,14.76.43,15.29a10.2,10.2,0,0,0,.16,1.5,5.72,5.72,0,0,0,.33,1.34c0.14,0.41.26,0.82,0.42,1.19,0.37,0.71.67,1.38,1,1.94l1,1.46c0.32,0.41.63,0.75,0.87,1s0.51,0.09.43-.22-0.23-.75-0.35-1.23L4,20.69c-0.1-.58-0.09-1.22-0.14-1.86,0-.32.05-0.65,0.08-1a3.44,3.44,0,0,1,.16-1A6.44,6.44,0,0,1,4.41,16l0.41-.8c0.2-.22.38-0.44,0.55-0.65L6,14c0.23-.14.5-0.24,0.72-0.37a7.52,7.52,0,0,1,.79-0.25,4.48,4.48,0,0,1,.84-0.15l0.41-.06H9.22c0.3,0,.56,0,0.85,0l0.72,0.07a3.77,3.77,0,0,1,1.2.21v3.17a1.25,1.25,0,0,0,2,.89l9-7.45A1.46,1.46,0,0,0,23.06,7.83Z" style="fill:#aaa"/></svg>
```

`forward.svg`

```xml
<svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"><path d="M23.61,11.07L17.07,4.35A1.2,1.2,0,0,0,15,5.28V9H1.4A1.82,1.82,0,0,0,0,10.82v2.61A1.55,1.55,0,0,0,1.4,15H15v3.72a1.2,1.2,0,0,0,2.07.93l6.63-6.72A1.32,1.32,0,0,0,23.61,11.07Z" style="fill:#aaa"/></svg>
```

HeahDude commented:
Thanks @javiereguiluz, committed.

HeahDude commented on src/Symfony/Component/HttpKernel/DataCollector/RequestDataCollector.php:
```diff
@@ -122,48 +122,34 @@ public function collect(Request $request, Response $response, \Exception $except
+        if ($request->hasSession() && $response->isRedirect()) {
+            $request->getSession()->set('sf_redirect', array(
+                'token' => $response->headers->get('x-debug-token'),
+                'route' => $request->attributes->get('_route', 'n/a'),
```
what about collecting controller too and making the route in wdt info panel a link to it ?

HeahDude replied:
We could also catch here the redirection code if we want to display it in wdt.

HeahDude commented on src/Symfony/Component/HttpKernel/DataCollector/RequestDataCollector.php:
```diff
@@ -276,18 +262,41 @@ public function getRouteParams()
+
+        if ($parentRequestAttributes = $event->getRequest()->attributes->get('_forwarded')) {
+            if ($parentRequestAttributes instanceof ParameterBag) {
+                $parentRequestAttributes->set('_forward_controller', $this->parseController($event->getController()));
```
I don't think it's need to test if `parentRequest->has('_forward_controller')` before overriding it.

Forward action should be use to delegate the return of the response to another controller as it also duplicate the original request. What would be the use case for calling it after the $response has been set ? There is no point to duplicate original request twice IMO.

Though overriding `_forward_controller` value seems a logical behaviour, we could even through an exception from abstract controller in framework bundle like my example of Silex usage in the description.

HeahDude commented on src/Symfony/Component/HttpKernel/DataCollector/RequestDataCollector.php:
```diff
@@ -122,48 +122,34 @@ public function collect(Request $request, Response $response, \Exception $except
+
+        if ($request->hasSession() && $request->getSession()->has('sf_redirect')) {
+            $this->data['redirect'] = $request->getSession()->get('sf_redirect');
+            $request->getSession()->remove('sf_redirect');
```
In case of chained redirections $response, each request collects its own previous $request profile, it will then also appear in wdt while on profiler intercept redirection page.

![chained_redirect](https://cloud.githubusercontent.com/assets/10107633/12719686/060dabb2-c8f7-11e5-8f1f-12cdd08f9fa0.jpg)

HeahDude commented:
@javiereguiluz WDYT of a redirect arrow pointing left instead of right, it would be less confusing and consistent with mail buttons definition for example.

Here's a screenshot from Mail app in OS X :

![mail](https://cloud.githubusercontent.com/assets/10107633/12734269/c9506f22-c93f-11e5-8ce8-b86b6621a615.jpg)

javiereguiluz commented:
@HeahDude I don't agree with the last proposal. The reason is that our toolbar displays "redirected to" and the mail app shows "redirected/replied from". The tip of the arrow should point to the route name because that route is the result of the redirection or forward.

* `200 -> @route` can be read as "a 200 response redirected to @route".
* `200 <- @route` can be read as "@route redirects to 200".

HeahDude commented:
@javiereguiluz alright. WDYT about collecting the redirected controller and/or redirect status code ? (see my notes)

HeahDude commented:
Hi @javiereguiluz, do I have to write some tests for this PR ?

Otherwise it is finished, I'm just waiting for some feedbacks on catching the redirection status code, and/or the route name which we've been redirected from.

Any lead ?

xabbuh commented on src/Symfony/Bundle/FrameworkBundle/Controller/Controller.php:
```diff
@@ -64,8 +64,10 @@ protected function generateUrl($route, $parameters = array(), $referenceType = U
     protected function forward($controller, array $path = array(), array $query = array())
     {
+        $request = $this->container->get('request_stack')->getCurrentRequest();
+        $path['_forwarded'] = $request->attributes;
```
I think this should be moved to an event listener instead. Otherwise you couldn't track subrequests triggered by user code that doesn't use the base controller class.

HeahDude replied:
@xabbuh, the point is not to catch subrequests as profiler already does, we can see them in the request panel.
This feature is about catching the "main" forward", using this special method, as it should be used only once in most use case to quickly click the profile token link of the original request in the wdt.

But I can try to go further, how would you imagine the presentation of sub requests token in the wdt ?

xabbuh commented:
What when there is more than one redirection before the WDT is shown again?

HeahDude commented:
@xabbuh for chained redirection see my [comment above](https://github.com/symfony/symfony/pull/17589#discussion_r51422405) but I could add a link in the top panel of the profiler for the case where `intercept_redirect` is false.

HeahDude commented:
So in that it could worth to catch controller and route info from the redirection response

HeahDude committed:
[WebProfilerBundle] [FrameworkBundle] profile forward controller action

closes #14358

The `Symfony\Bundle\FrameworkBundle\Controller\Controller::forward()`
now passes request attributes to its sub request, so when the profiler
will profile the sub request it will add its token to its hold by
reference parent request attributes.

The profiler main profiler layout displays a shortcut to the forwarded
sub request profile which is actually responsible for returning the
present response.

The web debug toolbar shows a notifying icon, meaning a shortcut to a
forwarded profile is available in the toolbar request panel.

HeahDude commented:
Ok, I've squashed things up !

So now there is two different commit for each feature :
-   3dc8844 for redirections profiling.
    
    In the scope of `RedirectResponse` sent by the kernel it will pass the request attributes to the next request through the session if it is enabled.

  It's seems consistent as the implementation globally remains in the `RequestDataCollector` from the HttpKernel component with requires HttpFoundation for request, response and session objects.

  I've refactored a bit since I've noticed that actually there is absolutely no information in the profiler when you are redirected unless you turn on interceptions (which is not by default), even with my previous implementation, excepted a residue of my collect deep in the session tab of the request panel.

  So I've added a little info (a simple block) in the summury in the main profiler layout, plus the status code in the wdt. Here's what it looks like :
  (the route name may be a link to the controller)
![redirect_wdt](https://cloud.githubusercontent.com/assets/10107633/13279850/b3dba8da-dada-11e5-9b95-c4b676941336.jpg)

  If you click on the linked profile token you can see it is a chained redirection :
![redirection_profile](https://cloud.githubusercontent.com/assets/10107633/13279958/6439ea3e-dadb-11e5-8398-3fd058d160c7.jpg)

- 79e5464 for a shorcut to the forward profile.

  Since this feature depends on the `Symfony\Bundle\FramworkBundle\Controller\Controller` it may not be consistent to put its logic in the `RequestDataCollector`, especially as it tests on every request if there is a forwarded attribute on the kernel controller event.

    May be I should add a dedicated listener like `\Symfony\Bundle\FrameworkBundle\DataCollector\ForwardRequestCollector`.

    Or maybe you don't want this feature after all, what do you think symfony deciders ?

@javiereguiluz, can you have a look please ? :)

fabpot commented:
@javiereguiluz Can you have a look at this one? If we are able to manage redirections directly from the toolbar, I suggest that we remove the setting for the intermediate page as it would not be that useful anymore.

javiereguiluz commented:
I'm gonna review this soon. @HeahDude sorry for the delay.

HeahDude commented:
For what it's worth, I don't think this is redundant with the interception page as in case of chain redirections there are complementary (ref https://github.com/symfony/symfony/pull/17589#discussion_r51422405).

As an option set false by default, it should be up to the developer to decide if he wants to use intercept page.

It can allow to check some dumps or profile data before dealing with the next request.

The wdt info is just a helper but may not be replacement for intercept option IMHO.

HeahDude commented:
@javiereguiluz no worry :)

HeahDude committed:
[FrameworkBundle] Extends the RequestDataCollector

Moves the logic of collect controller data while forwarding
in the framework bundle.

HeahDude commented:
Ok I've pushed a new commit 6b7991b, to move the forward catching logic in the FrameworkBundle.

That makes the tip for Silex in the description obsolete but I would be bad to check on each controller event if it a forward since it a framework controller specific method.

If you agree, I'll squash the alternative.

Beside that, I think it's done here.

Needs final review.

javiereguiluz commented:
I've reviewed this feature and it worked for me. @HeahDude thanks for this great work!

Now I'd like to propose some minor changes.

---

In the toolbar, redirects are displayed like this:

![before_toolbar_redirect](https://cloud.githubusercontent.com/assets/73419/13462505/08ff73c0-e088-11e5-9a4f-2bb55390b795.png)

I'd like to separate the redirection information and to reword its contents:

![after_toolbar_redirect](https://cloud.githubusercontent.com/assets/73419/13462813/8e623bfa-e089-11e5-8556-b7e286d5731c.png)

Same idea for forwards:

![after_toolbar_forward](https://cloud.githubusercontent.com/assets/73419/13462525/2bfe0e40-e088-11e5-8ef5-cc0a994cf037.png)

---

In the profiler, the redirection information is displayed at the end of the second line in the green section:

![before_profiler_redirect](https://cloud.githubusercontent.com/assets/73419/13462556/4462fd92-e088-11e5-9e76-5b8a2c13e3e2.png)

Since this information can be very important, I'd like to show it more prominently:

![after_profiler_redirect](https://cloud.githubusercontent.com/assets/73419/13462569/58596980-e088-11e5-9eed-b0c0b41bf80e.png)

And the same idea for forward:

![after_profiler_forward](https://cloud.githubusercontent.com/assets/73419/13462581/5e8d3f8e-e088-11e5-93ea-d803bcf5a0c8.png)

---

I attach you a patch with these changes in case Symfony Deciders agree with them:

```diff
diff --git a/src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/request.html.twig b/src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/request.html.twig
index 3be795e..58fcac9 100644
--- a/src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/request.html.twig
+++ b/src/Symfony/Bundle/WebProfilerBundle/Resources/views/Collector/request.html.twig
@@ -25,60 +25,68 @@
     {% set icon %}
         <span class="sf-toolbar-status sf-toolbar-status-{{ request_status_code_color }}">{{ collector.statuscode }}</span>
         {% if collector.route %}
-            {% if collector.redirect %}&nbsp;{{ include('@WebProfiler/Icon/redirect.svg') }}{% endif %}
-            {% if collector.forward %}&nbsp;{{ include('@WebProfiler/Icon/forward.svg') }}{% endif %}
+            {% if collector.redirect %}{{ include('@WebProfiler/Icon/redirect.svg') }}{% endif %}
+            {% if collector.forward %}{{ include('@WebProfiler/Icon/forward.svg') }}{% endif %}
             <span class="sf-toolbar-label">@</span>
             <span class="sf-toolbar-value sf-toolbar-info-piece-additional">{{ collector.route }}</span>
         {% endif %}
     {% endset %}
 
     {% set text %}
-        <div class="sf-toolbar-info-piece">
-            <b>HTTP status</b>
-            <span>{{ collector.statuscode }} {{ collector.statustext }}</span>
-        </div>
-
-        <div class="sf-toolbar-info-piece">
-            <b>Controller</b>
-            <span>{{ request_handler }}</span>
-        </div>
+        <div class="sf-toolbar-info-group">
+            <div class="sf-toolbar-info-piece">
+                <b>HTTP status</b>
+                <span>{{ collector.statuscode }} {{ collector.statustext }}</span>
+            </div>
 
-        {% if collector.controller.class is defined %}
             <div class="sf-toolbar-info-piece">
-                <b>Controller class</b>
-                <span>{{ collector.controller.class }}</span>
+                <b>Controller</b>
+                <span>{{ request_handler }}</span>
             </div>
-        {% endif %}
 
-        <div class="sf-toolbar-info-piece">
-            <b>Route name</b>
-            <span>{{ collector.route|default('NONE') }}</span>
-        </div>
+            {% if collector.controller.class is defined %}
+                <div class="sf-toolbar-info-piece">
+                    <b>Controller class</b>
+                    <span>{{ collector.controller.class }}</span>
+                </div>
+            {% endif %}
 
-        <div class="sf-toolbar-info-piece">
-            <b>Has session</b>
-            <span>{% if collector.sessionmetadata|length %}yes{% else %}no{% endif %}</span>
-        </div>
+            <div class="sf-toolbar-info-piece">
+                <b>Route name</b>
+                <span>{{ collector.route|default('NONE') }}</span>
+            </div>
 
-        {% if collector.redirect %}
             <div class="sf-toolbar-info-piece">
-            <b>Redirected from</b>
-            <span>
-                <span class="sf-toolbar-redirection-status sf-toolbar-status-yellow">{{ collector.redirect.status_code }}</span>
-                {{ redirect_handler }}
-                (<a href="{{ path('_profiler', { token: collector.redirect.token }) }}">{{ collector.redirect.token }}</a>)
-            </span>
+                <b>Has session</b>
+                <span>{% if collector.sessionmetadata|length %}yes{% else %}no{% endif %}</span>
+            </div>
         </div>
+
+        {% if collector.redirect %}
+            <div class="sf-toolbar-info-group">
+                <div class="sf-toolbar-info-piece">
+                    <b>
+                        <span class="sf-toolbar-redirection-status sf-toolbar-status-yellow">{{ collector.redirect.status_code }}</span>
+                        Redirect from
+                    </b>
+                    <span>
+                        {{ redirect_handler }}
+                        (<a href="{{ path('_profiler', { token: collector.redirect.token }) }}">{{ collector.redirect.token }}</a>)
+                    </span>
+                </div>
+            </div>
         {% endif %}
 
         {% if forward_handler is defined %}
-            <div class="sf-toolbar-info-piece">
-            <b>Forwarded to</b>
-            <span>
-                {{ forward_handler }}
-                (<a href="{{ path('_profiler', { token: collector.forward.token }) }}">{{ collector.forward.token }}</a>)
-            </span>
-        </div>
+            <div class="sf-toolbar-info-group">
+                <div class="sf-toolbar-info-piece">
+                    <b>Forwarded to</b>
+                    <span>
+                        {{ forward_handler }}
+                        (<a href="{{ path('_profiler', { token: collector.forward.token }) }}">{{ collector.forward.token }}</a>)
+                    </span>
+                </div>
+            </div>
         {% endif %}
     {% endset %}
 
diff --git a/src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/layout.html.twig b/src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/layout.html.twig
index 7518b49..3d3d423 100644
--- a/src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/layout.html.twig
+++ b/src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/layout.html.twig
@@ -19,6 +19,44 @@
                             {% endif %}
                         </h2>
 
+                        {% if profile.collectors.request is defined and profile.collectors.request.redirect %}
+                            {% set redirect = profile.collectors.request.redirect %}
+                            {% set controller = redirect.controller %}
+                            <dl class="metadata">
+                                <dt>
+                                    <span class="label">{{ redirect.status_code }}</span>
+                                    Redirect from
+                                </dt>
+                                <dd>
+                                    {% if redirect.controller.class is defined %}
+                                        {% set link = controller.file|file_link(controller.line) %}
+                                        {% if link %}<a href="{{ link }}" title="{{ controller.file }}">{% endif %}
+                                            @{{ redirect.route }}
+                                        {% if link %}</a>{% endif %}
+                                    {% else %}
+                                            @{{ redirect.route }}
+                                    {% endif %}
+                                    (<a href="{{ path('_profiler', { token: redirect.token }) }}">{{ redirect.token }}</a>)
+                                </dd>
+                            </dl>
+                        {% endif %}
+
+                        {% if profile.collectors.request is defined and profile.collectors.request.forward and profile.collectors.request.forward.controller.class is defined %}
+                            {% set forward = profile.collectors.request.forward %}
+                            {% set controller = forward.controller %}
+                            <dl class="metadata">
+                                <dt>Forwarded to</dt>
+                                <dd>
+                                    {% set link = controller.file|file_link(controller.line) %}
+                                    {% if link %}<a href="{{ link }}" title="{{ controller.file }}">{% endif %}
+                                        {{- controller.class|abbr_class|striptags -}}
+                                        {{- controller.method ? ' :: ' ~ controller.method -}}
+                                    {% if link %}</a>{% endif %}
+                                    (<a href="{{ path('_profiler', { token: forward.token }) }}">{{ forward.token }}</a>)
+                                </dd>
+                            </dl>
+                        {% endif %}
+
                         <dl class="metadata">
                             <dt>Method</dt>
                             <dd>{{ profile.method|upper }}</dd>
@@ -34,26 +72,6 @@
 
                             <dt>Token</dt>
                             <dd>{{ profile.token }}</dd>
-
-                            {% if profile.collectors.request is defined and profile.collectors.request.redirect -%}
-                            {%- set redirect = profile.collectors.request.redirect -%}
-                                <dt>Redirected from</dt>
-                                <dd>
-                                    {% set controller = redirect.controller -%}
-                                    {%- if redirect.controller.class is defined -%}
-                                    {%- set link = controller.file|file_link(controller.line) -%}
-                                        {% if link %}<a href="{{ link }}" title="{{ controller.file }}">{% endif -%}
-
-                                        @{{- redirect.route -}}
-
-                                        {%- if link %}</a>{% endif -%}
-                                    {%- else -%}
-                                            @{{ redirect.route }}
-                                    {%- endif %}
-                                    - {{ redirect.status_code }}&nbsp;{{ redirect.status_text }}
-                                    (<a href="{{ path('_profiler', { token: redirect.token }) }}">{{ redirect.token }}</a>)
-                                </dd>
-                            {%- endif %}
                         </dl>
                     </div>
                 </div>
diff --git a/src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/profiler.css.twig b/src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/profiler.css.twig
index 21e74b9..7621fd3 100644
--- a/src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/profiler.css.twig
+++ b/src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/profiler.css.twig
@@ -503,6 +503,9 @@ tr.status-warning td {
 #summary dl.metadata dd {
     margin: 0 1.5em 0 0;
 }
+#summary dl.metadata .label {
+    background: rgba(255, 255, 255, 0.2);
+}
 
 {# Sidebar
    ========================================================================= #}
diff --git a/src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/toolbar.css.twig b/src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/toolbar.css.twig
index 8cd00f6..218cb32 100644
--- a/src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/toolbar.css.twig
+++ b/src/Symfony/Bundle/WebProfilerBundle/Resources/views/Profiler/toolbar.css.twig
@@ -233,7 +233,8 @@
     text-decoration: underline;
 }
 .sf-toolbar-block-request .sf-toolbar-redirection-status {
-    padding: 0 4px;
+    font-weight: normal;
+    padding: 2px 4px;
     line-height: 18px;
 }
 
@@ -453,8 +454,11 @@
         padding-left: 0;
         padding-right: 0;
     }
-    .sf-toolbar-block-request .sf-toolbar-status + .sf-toolbar-label {
-        margin-left: 4px;
+    .sf-toolbar-block-request .sf-toolbar-status {
+        margin-right: 5px;
+    }
+    .sf-toolbar-block-request .sf-toolbar-icon svg + .sf-toolbar-label {
+        margin-left: 0;
     }
     .sf-toolbar-block-request .sf-toolbar-label + .sf-toolbar-value {
         margin-right: 10px;
```

fabpot commented:
@javiereguiluz looks good to me.

HeahDude commented:
Thanks @javiereguiluz for the review, I work on it.

@fabpot what about https://github.com/symfony/symfony/pull/17589#issuecomment-191144427 was it part of your last agreement ?

HeahDude commented:
@javiereguiluz, I propose to either use "Redirected from" and  "Forwarded to", or "Redirect from" and "Forward to", (I vote for the passive form) but a mix of both looks weird to me.

What do you think ?

javiereguiluz commented:
@HeahDude I know that they differ ... but I proposed that on purpose. It's part of the "conversational UI" trend. A person speaking in English would always say *"a permanent redirect to/from ..."* or *"a temporary redirect to/from..."*.

That's what the interface is trying to do in this screen:

![after_toolbar_redirect](https://cloud.githubusercontent.com/assets/73419/13471319/561fd672-e0b0-11e5-9844-7c27be3bdafa.png)

And this also allows us to align the two different HTTP status, which improves the overall readability.

HeahDude commented:
@javiereguiluz thank you very much for explaining this to me, as a french php self-learner, it's a kind of nuance very useful to learn :)

dkisselev commented:
@HeahDude, English is difficult that way, I'm a native speaker and still had to stop and think about the correct phrasing when you mentioned that. I agree with @javiereguiluz on conversational UI though.

HeahDude commented:
Ok I'm almost done addressing last @javiereguiluz comment. Just need to know if I discard or squash 6b7991b alternative.

ping @javiereguiluz @fabpot

HeahDude committed:
[WebProfilerBundle] show the http method in wdt if not 'GET'

When the request method is different from 'GET', the wdt add a
hint before the route name in the request panel.

HeahDude commented:
Ok, I've rebased. Here's a new summary :

1. 04d4dfc Handle redirection profiling
  last @javiereguiluz's comment addressed.
  Also I've tweaked the macro to be consistent with the proposed diff (see [here](https://github.com/symfony/symfony/pull/17589/files#diff-42750072126a9fed4fad2f43c4b73d03R277)).

2. 23a7106 A new proposal to show the redirected method when different from 'GET'
  Kernel redirections are often triggered to change the http verb, so it might be useful to show it when needed. Here's a glimpse :

  ![redirect_method_hint](https://cloud.githubusercontent.com/assets/10107633/13491458/bf2cc36e-e132-11e5-84c5-358895a916ba.jpg)
  ![hint_method_redirect_layout](https://cloud.githubusercontent.com/assets/10107633/13491508/01c9dc3e-e133-11e5-909a-378c85eac31e.jpg)

3. 5f1cc2e Is a patch for intercept_redirect` option compatibility.

  ------------------
  >Form here if accepted, 2 & 3 should be squashed in 1.

  -----------------

4. 1218e84 Forward profiling
  As it has been added in profiler main layout too, I've set a view var `request_collector` to ease reading.
  `{% if profile.collectors.request is defined and profile.collectors.request.forward and profile.collectors.request.forward.controller.class is defined %}`
  This is crazy!

5. 2653b64 Separation of concerns, maybe discard or squashed in 4.

  There is a point to discard 6b7991b imo, because keeping the logic in http-foundation allows to catch global custom forward implementation.

  If the sub request build in some controller is not only a fragment but the return statement from a controller, this should be considered as a forward action.
  So to specifically profile this sub request, one just have to set it an attribute with the key "_forwarded" and the profiler collects the controller, route and profile token from it.

  Actually it does pretty much the same thing as for the redirection feature.

  The FrameworkBundle and the WebProfilerBundle may just use one of many possible implementations of a forward profiling feature for a base controller class.
  This may have even more sense if the HttpProfiler had its own component (ref #.

  My example in the [description](#) of this PR with Silex shows a supported implementation, but this method could be in a trait of any app or base controller class.

  Then it should be part of the profiler dx to support this option out of the box.

  Otherwise it's the framework bundle only responsibility to handle the forward feature, hence I would squash it as I think that anyone willing to implement a custom forward feature pattern should extend the `RequestDataCollector` to profile it.

6. 9ba4a1a My last proposal

  IIRC in the "old" profiler the route was displayed in the wdt with the http method (`GET@route_name`).
  Like in point 2, I think it would be useful to show it only when the method is different form 'GET'.
  Here's a glimpse :
  ![method_hint](https://cloud.githubusercontent.com/assets/10107633/13491876/322b9690-e135-11e5-9780-4548336e93e3.jpg)

javiereguiluz commented:
@HeahDude thanks for the updates. I will review everything soon, but I have a quick comment. The `@` in `@a_string` syntax is used to denote that the string is in fact a route name. I wouldn't add the HTTP method information to the `@` character.

If this info is needed, we could show it like this:

![patch_method](https://cloud.githubusercontent.com/assets/73419/13492243/4b8fbb82-e137-11e5-9d7e-54a156f16b84.png)

HeahDude commented:
Ok @javiereguiluz, I took care of your last comment and updated my [summary](https://github.com/symfony/symfony/pull/17589#issuecomment-191703462). Ready for review.

You should try chained redirects and/with the interception page.

Maybe you'll have some ideas to enhance it or agree with Fabien's https://github.com/symfony/symfony/pull/17589#issuecomment-190285961

javiereguiluz commented:
I've tested everything again and it works as expected:

### Redirect in the toolbar

![normal_redirect_toolbar](https://cloud.githubusercontent.com/assets/73419/14135229/a9ad63b0-f65c-11e5-992a-8ccb11fa0125.png)

### Redirect in the profiler

![normal_redirect_profiler](https://cloud.githubusercontent.com/assets/73419/14135240/adc2eb5a-f65c-11e5-8e9a-10974bb080ec.png)

### Intercepted Redirect

![captured_redirect](https://cloud.githubusercontent.com/assets/73419/14135243/b158f520-f65c-11e5-8184-15b7bd5673b8.png)

### Forward in the toolbar

![forward_toolbar](https://cloud.githubusercontent.com/assets/73419/14135247/b588e5ba-f65c-11e5-89c7-7b209fed088d.png)

### Forward in the profiler

![forward_profiler](https://cloud.githubusercontent.com/assets/73419/14135254/bfe1d09e-f65c-11e5-8086-41455a2282cd.png)

---

@HeahDude do you think you could take care of the test fails before tomorrow's feature freeze? I'd like Symfony 3.1 to include this great feature. Thanks!

HeahDude commented:
@javiereguiluz Thanks very much for the review :)

I'll take a look before this evening.

I've still have concerns about the refactoring of 5073f30, and should I make only one commit in your opinion ?

There is actually 3 commits holding a single feature: redirect, forward and method hints.

cc/ @fabpot

HeahDude commented:
Oops, missed this [test](https://travis-ci.org/symfony/symfony/jobs/113912575#L2570) is that a BC break ? :fearful:

fabpot commented:
There is no BC promise on the data collector by the profiler from one version to the next. Profiles are for debugging only, so format compatibility is not needed.

xabbuh commented:
The tests need to be fixed though.

HeahDude commented:
Cool!

HeahDude commented:
It should be ok now. Actually I had to change that test because I did not just move the controller resolution in a private method, I also add these [3 lines](https://github.com/symfony/symfony/pull/17589/commits/0a1b2841f7ca964551c8339ef031ae2fb84c50b5#diff-e8f5b14fbfbbeac60fc9f3abe310c3b0R342) to help dealing with both redirection and forward controllers' data.

@javiereguiluz In your last review it looks like you missed to test method hint. Could you please try with a simple post / post @ redirect. Thank you very much for your leading on this PR :)

Please any reviewers, try to look at this PR by commit, they all have a summary block.

ping __@symfony/deciders__

javiereguiluz commented:
This feature looks ready to be merged into 3.1 before the "feature freeze".

HeahDude commented:
:) So there is no squashing needed here ? 4f020b5 could be "fixup" in "227ac77" if we intend to keep it.

fabpot commented:
@HeahDude If you want to rearrange the commits before merge, please do so.

HeahDude commented:
@fabpot The question is, is it needed to keep them separately ?

fabpot commented:
I don't mind keeping several commits if they make sense and it seems they do :)

HeahDude commented:
Agreed :)

So we're done here :tada:

fabpot commented:
Thank you @HeahDude.

HeahDude commented:
I would like to thank you and all the community for sharing symfony :)

I've learned some much thanks to you all since I use it, merci!

rvanlaak commented:
Missed the development process of this feature, but it looks awesome @HeahDude ! :rocket:
